### PR TITLE
feature/user-select-llm

### PR DESCRIPTION
**Pull request by gecBurton**
## Context

As an admin user I want to be able to change the LLM we are using for chat from the frontend

## Changes proposed in this pull request

1. `chat_backend` has been moved from `Settings` to `AISettings`
2. `Settings` now contains 3 sets of model-names, api-keys, and endpoints, the  `AISettings` is now used to dynamically choose swap between them
3. rather than call `get_llm(env)` on the initialisation of `Redbox` `get_chat_llm(env, ai_settings)` is now called where needed from within the nodes
4. rather than using dependency injection to set the llm in the tests we now use monkey-patching to override `get_chat_llm`

## Relevant links
https://technologyprogramme.atlassian.net/browse/REDBOX-646

## Things to check

- [ ] I have added any new ENV vars in all deployed environments
- [ ] I have tested any code added or changed
- [ ] I have run integration tests https://github.com/i-dot-ai/redbox/actions/runs/10383971164

### CODE REVIEW

File: redbox-core/redbox/chains/components.py [hunk 1]
```diff
@@ -6,17 +7,42 @@
...
 
 
-def get_chat_llm(env: Settings):
-    return AzureChatOpenAI(
-        api_key=convert_to_secret_str(env.azure_openai_api_key),
-        azure_endpoint=env.azure_openai_endpoint,
-        model=env.azure_openai_model,
-        api_version=env.azure_api_version_embeddings,
-    )
+def get_chat_llm(env: Settings, ai_settings: AISettings):
```
Comment (gecBurton) [inline redbox-core/redbox/chains/components.py:15]: this is the key change

File: core-api/core_api/dependencies.py [hunk 2]
```diff
@@ -53,55 +49,9 @@ def get_all_chunks_retriever(env: Annotated[Settings, Depends(get_env)]):
     )
 
 
-@lru_cache(1)
-def get_llm(env: Annotated[Settings, Depends(get_env)]) -> ChatLiteLLM:
```
Comment (gecBurton) [inline core-api/core_api/dependencies.py:57]: now using the `get_chat_llm` instead

File: core-api/tests/routes/test_chat.py [hunk 3]
```diff
@@ -141,9 +141,6 @@ def test_case(request):
 @pytest.fixture
 def client(test_case: RedboxChatTestCase, embedding_model):
     chat_app.dependency_overrides[dependencies.get_embedding_model] = lambda: embedding_model
-    chat_app.dependency_overrides[dependencies.get_llm] = lambda: GenericFakeChatModel(
```
Comment (gecBurton) [inline core-api/tests/routes/test_chat.py:144]: using monkey patching instead

File: redbox-core/redbox/models/settings.py [hunk 4]
```diff
@@ -37,15 +37,26 @@ class ElasticCloudSettings(BaseModel):
 class Settings(BaseSettings):
     """Settings for the redbox application."""
 
-    anthropic_api_key: str | None = None
-    openai_api_key: str = "NotAKey"
-    azure_openai_api_key: str = "NotAKey"
-    azure_openai_endpoint: str | None = None
-
-    openai_api_version: str = "2023-12-01-preview"
+    # azure/gpt-35-turbo-16k
```
Comment (gecBurton) [inline redbox-core/redbox/models/settings.py:40]: we could move `openai_api_version_*`, `azure_openai_api_key_*` & `azure_openai_endpoint_*` to the AISettings in future

File: redbox-core/redbox/models/settings.py [hunk 5]
```diff
@@ -37,15 +37,26 @@ class ElasticCloudSettings(BaseModel):
...
+    # azure/gpt-4
+    openai_api_version_4t: str = "2024-02-01"
+    azure_openai_api_key_4t: str = "not a key"
+    azure_openai_endpoint_4t: str = "not an endpoint"
+
+    # azure/gpt-4o
+    openai_api_version_4o: str = "2024-02-01"
+    azure_openai_api_key_4o: str = "not a key"
+    azure_openai_endpoint_4o: str = "not an endpoint"
+
```
Comment (gecBurton) [inline redbox-core/redbox/models/settings.py:54]: `embedding_openai_api_key` is specifically for the Embeddings ad will match one of:
* `azure_openai_api_key_35t`
* `azure_openai_api_key_4t`
* `azure_openai_api_key_4o`

File: .github/workflows/integration.yml [hunk 6]
```diff
@@ -63,9 +63,10 @@ jobs:
         mkdir -p data/elastic/
         chmod -R 777 data/
         cp .env.integration .env
-        echo AZURE_OPENAI_ENDPOINT=${{ secrets.AZURE_OPENAI_ENDPOINT }} >> .env
-        echo AZURE_OPENAI_API_KEY=${{ secrets.AZURE_OPENAI_API_KEY }} >> .env
-        echo AZURE_OPENAI_MODEL=azure/gpt-4-turbo-2024-04-09 >> .env
+        echo AZURE_OPENAI_ENDPOINT_4T=${{ secrets.AZURE_OPENAI_ENDPOINT }} >> .env
+        echo AZURE_OPENAI_API_KEY_4T=${{ secrets.AZURE_OPENAI_API_KEY }} >> .env
+        echo EMBEDDING_AZURE_OPENAI_ENDPOINT=${{ secrets.AZURE_OPENAI_ENDPOINT }} >> .env
+        echo EMBEDDING_OPENAI_API_KEY=${{ secrets.AZURE_OPENAI_API_KEY }} >> .env
```
Comment (gecBurton) [inline .github/workflows/integration.yml:69]: we need to set the same variables twice: once for the embedding model and once for the relevant chat model

File: django_app/redbox_app/redbox_core/models.py [hunk 7]
```diff
@@ -46,6 +46,11 @@ def sanitise_string(string: str | None) -> str | None:
 
 
 class AISettings(UUIDPrimaryKeyBase, TimeStampedModel):
+    class ChatBackend(models.TextChoices):
+        GPT_35_TURBO = "gpt-35-turbo-16k", _("gpt-35-turbo-16k")
+        GPT_4_TURBO = "gpt-4-turbo-2024-04-09", _("gpt-4-turbo-2024-04-09")
+        GPT_35_OMNI = "gpt-4o", _("gpt-4o")
```
Comment (lmwilkigov) [inline django_app/redbox_app/redbox_core/models.py:52]: This Enum value for GPT-4o looks wrong

File: redbox-core/redbox/chains/components.py (hunk 1 above)
Comment (lmwilkigov) [reply redbox-core/redbox/chains/components.py:15]: Very clean, but should we document somewhere that the the default model will be GPT 3.5 turbo if that is that key is present in the .env over any other defined models? Just conscious that busy envs in future might cause headaches

**Review by lmwilkigov — CHANGES_REQUESTED** [review]

File: django_app/redbox_app/redbox_core/models.py (hunk 7 above)
Comment (gecBurton) [reply django_app/redbox_app/redbox_core/models.py:52]: good spot!

Comment (gecBurton) [reply django_app/redbox_app/redbox_core/models.py]: [ee26cea](https://github.com/i-dot-ai/redbox/pull/958/commits/ee26ceaa8596cf8b6314a03d1babdf0853190f8b)

**Comment by gecBurton** [comment]
@lmwilkigov I have added an explanation here https://github.com/i-dot-ai/redbox/pull/958#discussion_r1716548890

**Review by lmwilkigov — APPROVED** [review]

**Pull request merged**